libraries/AP_CheckFirmware/AP_CheckFirmware.cpp

```diff
@@ -47,7 +78,54 @@ check_fw_result_t check_good_firmware(void)
     if (crc1 != ad->image_crc1 || crc2 != ad->image_crc2) {
         return check_fw_result_t::FAIL_REASON_BAD_CRC;
     }
-    return check_fw_result_t::CHECK_FW_OK;
+
+    check_fw_result_t ret = check_fw_result_t::CHECK_FW_OK;
+
+#if AP_SIGNED_FIRMWARE
+    uint8_t hash[32];
+    int err = wc_ecc_init(&publicKey);
+    if (err != MP_OKAY) {
+        wc_ecc_free(&publicKey);
+        uprintf("Failed to init public key in bootloader\r\n");
+        return check_fw_result_t::FAIL_REASON_BAD_PUBLIC_KEY;
+    }
+
+    // verify signature
+    wc_InitSha256(&sha);
+    wc_Sha256Update(&sha, flash, len1);
+    wc_Sha256Update(&sha, (const uint8_t *)&ad->version_major, len2);
+    wc_Sha256Final(&sha, hash);
+
+    // do ecc verify
+    for (auto public_key : public_keys.public_key) {
```

Comment:
It seems like code loops through the keys until it finds a valid key breaking out of the loop if "passed". That's fine for the happy path, but if I understand this correctly, the implication is that if there are multiple keys and if it loops through all of them without finding a valid key, the method will only return the fail reason for the last key.
It might be more helpful to return an array of keys and failure reasons - for anyone trying to figure out why something isn't working.